### PR TITLE
Inject push-down filter to exclude indexed rows from deleted files

### PR DESCRIPTION
**Pull request by sezruby**
### What is the context for this pull request?

 - **Tracking Issue**: #150
 - **Parent Issue**: #150
 - **Dependencies**: #165

### What changes were proposed in this pull request?

With lineage column in the index data, we could exclude the deleted rows from the index data at query time, by injecting `Filter(Not(In(<lineage column>, <deleted files>)))`.
I added additional config `spark.hyperspace.index.hybridscan.delete.enabled` for enabling/disabling hybrid scan for delete dataset until performance validation and optimization.
- `spark.hyperspace.index.hybridscan.enabled = true/ spark.hyperspace.index.hybridscan.delete.enabled = false` => hybrid scan only for appended dataset; if there are deleted files in source relation, the index won't be applied.
- `spark.hyperspace.index.hybridscan.enabled = false / spark.hyperspace.index.hybridscan.delete.enabled = true` won't work

To support this, additional project & filter node is injected to LogicalRelation plan of index data. Reading the index data with the lineage column and filter the deleted files, and then exclude the lineage column as it shouldn't be shown in the result.
```
   :  :- Project [clicks#456, query#454]
   :  :  +- Filter ((isnotnull(clicks#456) && (clicks#456 >= 2000)) && (clicks#456 <= 4000))
   :  :     +- Project [Query#454, clicks#456]
   :  :        +- Filter NOT _data_file_name#474 IN  (file:/C:/Users/eunsong/repo/hyperspace2/src/test/resources/data/sampleparquet4/part-00003-3b411cdd-be12-4dad-a6a0-23fd244bd6e4-c000.snappy.parquet)
```

Note that the injected Project & Fitler is optimized by Spark so that there's no extra stage for them:
```
   :     +- BucketUnion 200 buckets, bucket columns: [clicks]
   :        :- *(1) Project [clicks#456, query#454]
   :        :  +- *(1) Filter ((((isnotnull(_data_file_name#474) && NOT (_data_file_name#474 = file:/C:/Users/eunsong/repo/hyperspace2/src/test/resources/data/sampleparquet4/part-00003-3b411cdd-be12-4dad-a6a0-23fd244bd6e4-c000.snappy.parquet)) && isnotnull(clicks#456)) && (clicks#456 >= 2000)) && (clicks#456 <= 4000))
   :        :     +- *(1) FileScan parquet [clicks#456,Query#454,_data_file_name#474] Batched: true, Format: Parquet, Location: InMemoryFileIndex[file:/C:/Users/eunsong/repo/hyperspace2/src/test/resources/hybridScanTest/index..., PartitionFilters: [], PushedFilters: [IsNotNull(_data_file_name), Not(EqualTo(_data_file_name,file:/C:/Users/eunsong/repo/hyperspace2/..., ReadSchema: struct<clicks:int,Query:string,_data_file_name:string>, Sele
ctedBucketsCount: 200 out of 200
```

### Does this PR introduce _any_ user-facing change?

Yes, if a user enabled Hybrid Scan delete support, indexes invalidated by "deleted files" from its source relation can be still utilized with modified plan.

The followings are the logical & physical plan for Join plan, with both appended and deleted files.

Logical plan
```
Project [clicks#456, query#454, query#468]
+- Join Inner, (clicks#456 = clicks#470)
   :- BucketUnion 200 buckets, bucket columns: [clicks]
   :  :- Project [clicks#456, query#454]
   :  :  +- Filter ((isnotnull(clicks#456) && (clicks#456 >= 2000)) && (clicks#456 <= 4000))
   :  :     +- Project [Query#454, clicks#456]
   :  :        +- Filter NOT _data_file_name#474 IN (file:/C:/Users/eunsong/repo/hyperspace2/src/test/resources/data/sampleparquet4/part-00003-3b411cdd-be12-4dad-a6a0-23fd244bd6e4-c000.snappy.parquet)
   :  :           +- Relation[Query#454,clicks#456,_data_file_name#474] parquet
   :  +- RepartitionByExpression [clicks#456], 200
   :     +- Project [clicks#456, query#454]
   :        +- Filter ((isnotnull(clicks#456) && (clicks#456 >= 2000)) && (clicks#456 <= 4000))
   :           +- Relation[Query#454,clicks#456] parquet
   +- BucketUnion 200 buckets, bucket columns: [clicks]
      :- Project [clicks#470, query#468]
      :  +- Filter ((isnotnull(clicks#470) && (clicks#470 <= 4000)) && (clicks#470 >= 2000))
      :     +- Project [Query#468, clicks#470]
      :        +- Filter NOT _data_file_name#475 IN (file:/C:/Users/eunsong/repo/hyperspace2/src/test/resources/data/sampleparquet4/part-00003-3b411cdd-be12-4dad-a6a0-23fd244bd6e4-c000.snappy.parquet)
      :           +- Relation[Query#468,clicks#470,_data_file_name#475] parquet
      +- RepartitionByExpression [clicks#470], 200
         +- Project [clicks#470, query#468]
            +- Filter ((isnotnull(clicks#470) && (clicks#470 <= 4000)) && (clicks#470 >= 2000))
               +- Relation[Query#468,clicks#470] parquet
```

Physical plan
```
*(7) Project [clicks#456, query#454, query#468]
+- *(7) SortMergeJoin [clicks#456], [clicks#470], Inner
   :- *(3) Sort [clicks#456 ASC NULLS FIRST], false, 0
   :  +- *(3) Filter isnotnull(clicks#456)
   :     +- BucketUnion 200 buckets, bucket columns: [clicks]
   :        :- *(1) Project [clicks#456, query#454]
   :        :  +- *(1) Filter ((((isnotnull(_data_file_name#474) && NOT (_data_file_name#474 = file:/C:/Users/eunsong/repo/hyperspace2/src/test/resources/data/sampleparquet4/part-00003-3b411cdd-be12-4dad-a6a0-23fd244bd6e4-c000.snappy.parquet)) && isnotnull(clicks#456)) && (clicks#456 >= 2000)) && (clicks#456 <= 4000))
   :        :     +- *(1) FileScan parquet [clicks#456,Query#454,_data_file_name#474] Batched: true, Format: Parquet, Location: InMemoryFileIndex[file:/C:/Users/eunsong/repo/hyperspace2/src/test/resources/hybridScanTest/index..., PartitionFilters: [], PushedFilters: [IsNotNull(_data_file_name), Not(EqualTo(_data_file_name,file:/C:/Users/eunsong/repo/hyperspace2/..., ReadSchema: struct<clicks:int,Query:string,_data_file_name:string>, Sele
ctedBucketsCount: 200 out of 200
   :        +- Exchange hashpartitioning(clicks#456, 200)
   :           +- *(2) Project [clicks#456, query#454]
   :              +- *(2) Filter ((isnotnull(clicks#456) && (clicks#456 >= 2000)) && (clicks#456 <= 4000))
   :                 +- *(2) FileScan parquet [clicks#456,Query#454] Batched: true, Format: Parquet, Location: InMemoryFileIndex[file:/C:/Users/eunsong/repo/hyperspace2/src/test/resources/data/sampleparquet4/..., PartitionFilters: [], PushedFilters: [IsNotNull(clicks), GreaterThanOrEqual(clicks,2000), LessThanOrEqual(clicks,4000)], ReadSchema: struct<clicks:int,Query:string>
   +- *(6) Sort [clicks#470 ASC NULLS FIRST], false, 0
      +- *(6) Filter isnotnull(clicks#470)
         +- BucketUnion 200 buckets, bucket columns: [clicks]
            :- *(4) Project [clicks#470, query#468]
            :  +- *(4) Filter ((((isnotnull(_data_file_name#475) && NOT (_data_file_name#475 = file:/C:/Users/eunsong/repo/hyperspace2/src/test/resources/data/sampleparquet4/part-00003-3b411cdd-be12-4dad-a6a0-23fd244bd6e4-c000.snappy.parquet)) && isnotnull(clicks#470)) && (clicks#470 <= 4000)) && (clicks#470 >= 2000))
            :     +- *(4) FileScan parquet [clicks#470,Query#468,_data_file_name#475] Batched: true, Format: Parquet, Location: InMemoryFileIndex[file:/C:/Users/eunsong/repo/hyperspace2/src/test/resources/hybridScanTest/index..., PartitionFilters: [], PushedFilters: [IsNotNull(_data_file_name), Not(EqualTo(_data_file_name,file:/C:/Users/eunsong/repo/hyperspace2/..., ReadSchema: struct<clicks:int,Query:string,_data_file_name:string>, Sele
ctedBucketsCount: 200 out of 200
            +- ReusedExchange [clicks#470, query#468], Exchange hashpartitioning(clicks#456, 200)

```

### How was this patch tested?

unit test

### CODE REVIEW

File: src/main/scala/com/microsoft/hyperspace/index/rules/FilterIndexRule.scala [hunk 1]
```diff
@@ -147,12 +97,16 @@ object FilterIndexRule
       outputColumns: Seq[String],
       filterColumns: Seq[String],
       fsRelation: HadoopFsRelation): Seq[IndexLogEntry] = {
+    val indexManager = Hyperspace
+      .getContext(spark)
+      .indexCollectionManager
```
Comment (pirz) [inline src/main/scala/com/microsoft/hyperspace/index/rules/FilterIndexRule.scala:102]: I am just curious, is there a reason to move this code snippet outside of `Some` block. We dont use IndexManager for the `None` case.

File: src/main/scala/com/microsoft/hyperspace/index/rules/RuleUtils.scala [hunk 2]
```diff
@@ -106,4 +128,318 @@ object RuleUtils {
...
+   *
+   * NOTE: This method currently only supports replacement of Scan Nodes i.e. Logical relations
+   *
+   * @param spark Spark session.
+   * @param index Index used in replacement plan.
+   * @param plan Current logical plan.
+   * @param useBucketSpec Option whether to use BucketSpec for reading index data.
+   * @return Transformed logical plan that leverages an index.
+   */
+  private def transformPlanToUsePureIndexScan(
```
Comment (pirz) [inline src/main/scala/com/microsoft/hyperspace/index/rules/RuleUtils.scala:176]: (just a naming suggestion, feel free to ignore it if you disagree)
rename `transformPlanToUsePureIndexScan` -> `transformPlanToUseIndexOnlyScan`

File: src/main/scala/com/microsoft/hyperspace/index/rules/RuleUtils.scala [hunk 3]
```diff
@@ -106,4 +128,318 @@ object RuleUtils {
...
+      spark: SparkSession,
+      index: IndexLogEntry,
+      plan: LogicalPlan,
+      useBucketSpec: Boolean): LogicalPlan = {
+    val transformed = if (HyperspaceConf.hybridScanEnabled(spark)) {
+      transformPlanToUseHybridIndexDataScan(spark, index, plan, useBucketSpec)
+    } else {
+      transformPlanToUsePureIndexScan(spark, index, plan, useBucketSpec)
+    }
+    assert(!transformed.equals(plan))
```
Comment (pirz) [inline src/main/scala/com/microsoft/hyperspace/index/rules/RuleUtils.scala:159]: I think this `assert` should be replaced by an `exception` thrown.
Imagine the replacement fails and no transformation happens. I think, This assert basically makes the whole user query to break. One functionality aspect of Hyeprspace is that: "A user's query should not really break due to the "extra" optimization Hyperspace tries to add". That is why in the rules, we used `Exceptions` and in the `Catch` block of `transofrm` we catch any exception, log it and return the original plan. We dont really wanna fail a query if our attempt to (further) optimize fails.

File: src/main/scala/com/microsoft/hyperspace/index/rules/FilterIndexRule.scala (hunk 1 above)
Comment (sezruby) [reply src/main/scala/com/microsoft/hyperspace/index/rules/FilterIndexRule.scala:102]: I'll revert this in #165. Thanks :)

File: src/main/scala/com/microsoft/hyperspace/index/rules/RuleUtils.scala (hunk 2 above)
Comment (sezruby) [reply src/main/scala/com/microsoft/hyperspace/index/rules/RuleUtils.scala:176]: The name was also suggested.  I think `PureIndexScan` is also good naming :)

File: src/main/scala/com/microsoft/hyperspace/index/rules/RuleUtils.scala (hunk 3 above)
Comment (sezruby) [reply src/main/scala/com/microsoft/hyperspace/index/rules/RuleUtils.scala:159]: It's because there are pre-requite conditions and we previously check conditions for this like linear plan, candidate indexes, .. etc. I think some asserts are required for development & testing. If we only throw exceptions which are ignored in Catch block, it's hard to debug any issue or even unnoticed.
This assert was suggested by @imback82 in #165.

Comment (imback82) [reply src/main/scala/com/microsoft/hyperspace/index/rules/RuleUtils.scala:159]: Yes, assert makes more sense than exception here since we are validating the "assumption" for this function.

**Comment by imback82** [comment]
@sezruby Can you resolve the conflicts?

**Comment by imback82** [comment]
> * `spark.hyperspace.index.hybridscan.enabled = false / spark.hyperspace.index.hybridscan.delete.enabled = true` => hybrid scan only for appended dataset; if there are deleted files in source relation, the index won't be applied.
> * `spark.hyperspace.index.hybridscan.enabled = false / spark.hyperspace.index.hybridscan.delete.enabled = true` won't work

I think you meant `true` and `true` in the first bullet?

Btw, why do we need to introduce `spark.hyperspace.index.hybridscan.delete.enabled` config? Is there a particular use case we are targeting?

**Comment by sezruby** [comment]
@imback82 I added the config to make this delete support switchable until performance validation. However, we could keep both configs if needed.

Do you think we should also support delete-only case?

**Comment by imback82** [comment]
Ah, OK. We can revisit the config names once we have the perf numbers. Thanks! (It would have been nice if this info was captured in the PR description).

File: src/main/scala/com/microsoft/hyperspace/index/rules/RuleUtils.scala [hunk 4]
```diff
@@ -216,20 +223,35 @@ object RuleUtils {
     var unhandledAppendedFiles: Seq[Path] = Nil
 
     // Get transformed plan with index data and appended files if applicable.
-    val indexPlan = plan transformDown {
+    val indexPlan = plan transformUp {
+      // Use transformUp here as currently 1 Logical Relation is allowed (pre-requisite).
```
Comment (imback82) [inline src/main/scala/com/microsoft/hyperspace/index/rules/RuleUtils.scala:227]: nit: `1` -> `one`

File: src/main/scala/com/microsoft/hyperspace/index/rules/RuleUtils.scala [hunk 5]
```diff
@@ -216,20 +223,35 @@ object RuleUtils {
...
+      // Otherwise, we need additional check not to process newly injected Filter-Not-In nodes.
```
Comment (imback82) [inline src/main/scala/com/microsoft/hyperspace/index/rules/RuleUtils.scala:228]: Can we test the prerequisite as assert?
```scala
def verifyPrerequisite(plan: LogicalPlan) {
  getLogicalRelation(plan).isDefined
  // We can extend this as needed later.
}
```
And I think we can call `verifyPrerequisite` from `transformPlanToUseIndex` since its comment also has this as a pre-requisite.

File: src/main/scala/com/microsoft/hyperspace/index/rules/RuleUtils.scala [hunk 6]
```diff
@@ -216,20 +223,35 @@ object RuleUtils {
...
+          if (HyperspaceConf.hybridScanDeleteEnabled(spark) && index.hasLineageColumn(spark)) {
+            val (exist, nonExist) = curFileSet.partition(index.allSourceFileInfos.contains)
+            val filesAppended = nonExist.map(f => new Path(f.name))
+            if (exist.length < index.allSourceFileInfos.size) {
+              (index.allSourceFileInfos -- exist, filesAppended)
+            } else {
+              (Nil, filesAppended)
+            }
+          } else {
+            // Keep append-only implementation for efficiency.
```
Comment (imback82) [inline src/main/scala/com/microsoft/hyperspace/index/rules/RuleUtils.scala:247]: Let's remove `Keep`:

`// Append-only implementation of getting appended files for efficiency.`

Also, can you add comment that in `else`, there should be no deleted files, which is checked in `isHybridScanCandidate`? I think a better way is to keep a context to pass around. (we should do this as a follow up).

File: src/main/scala/com/microsoft/hyperspace/index/rules/RuleUtils.scala [hunk 7]
```diff
@@ -216,20 +223,35 @@ object RuleUtils {
...
+            (
+              Nil,
+              curFileSet.filterNot(index.allSourceFileInfos.contains).map(f => new Path(f.name)))
+          }
 
         val filesToRead = {
-          if (useBucketSpec || !isParquetSourceFormat) {
+          if (useBucketSpec || !isParquetSourceFormat || filesDeleted.nonEmpty) {
```
Comment (imback82) [inline src/main/scala/com/microsoft/hyperspace/index/rules/RuleUtils.scala:254]: Can you update the comment with `filesDeleted.nonEmpty` condition (and why)?

File: src/main/scala/com/microsoft/hyperspace/index/rules/RuleUtils.scala [hunk 8]
```diff
@@ -246,18 +268,33 @@ object RuleUtils {
           }
         }
 
+        val readSchema = if (filesDeleted.isEmpty) {
```
Comment (imback82) [inline src/main/scala/com/microsoft/hyperspace/index/rules/RuleUtils.scala:271]: `newSchema`? Also, can you add a comment why you have a different branch for delete (e.g., need to utilized DATA_FILE_NAME_COLUMN, etc.)? BTW, can you just append this data file name column to the `StructType(index.schema.filter(baseRelation.schema.contains(_)))` or no?

File: src/main/scala/com/microsoft/hyperspace/index/rules/RuleUtils.scala [hunk 9]
```diff
@@ -246,18 +268,33 @@ object RuleUtils {
...
           Map())(spark)
 
         val updatedOutput =
           baseOutput.filter(attr => relation.schema.fieldNames.contains(attr.name))
-        baseRelation.copy(relation = relation, output = updatedOutput)
+
+        if (filesDeleted.isEmpty) {
+          baseRelation.copy(relation = relation, output = updatedOutput)
+        } else {
+          val lAttr = AttributeReference(IndexConstants.DATA_FILE_NAME_COLUMN, StringType)(
```
Comment (imback82) [inline src/main/scala/com/microsoft/hyperspace/index/rules/RuleUtils.scala:292]: `AttributeReference(IndexConstants.DATA_FILE_NAME_COLUMN, StringType)()` (`NamedExpression.newExprId` is default).

also what is `l` in `lAttr`?

File: src/main/scala/com/microsoft/hyperspace/index/rules/RuleUtils.scala (hunk 5 above)
Comment (imback82) [reply src/main/scala/com/microsoft/hyperspace/index/rules/RuleUtils.scala:228]: Also, we only match `LogicalRelation` so does it matter if we do transform up or down? If I missed something can you elaborate in the comment?

File: src/test/scala/com/microsoft/hyperspace/index/rules/RuleUtilsTest.scala [hunk 10]
```diff
@@ -120,59 +120,94 @@ class RuleUtilsTest extends HyperspaceRuleTestSuite with SQLHelper {
...
         // Scenario #2: Delete 1 file.
-        {
-          FileUtils.delete(new Path(indexFile), isRecursive = false)
-        }
+        FileUtils.delete(new Path(sourceFile))
 
         {
           val optimizedPlan = spark.read.parquet(dataPath).queryExecution.optimizedPlan
-          verify(optimizedPlan, hybridScanEnabled = false, expectCandidateIndex = false)
-          // TODO: expectedCandidateIndex = true once delete dataset is supported.
```
Comment (imback82) [inline src/test/scala/com/microsoft/hyperspace/index/rules/RuleUtilsTest.scala:165]: Should we add `hybridScanDeleteEnabled = true` case?

File: src/test/scala/com/microsoft/hyperspace/index/HybridScanTest.scala [hunk 11]
```diff
@@ -39,22 +39,34 @@ class HybridScanTest extends QueryTest with HyperspaceSuite {
...
-    assert(appendCnt < inputFiles.length)
+    assert(appendCnt + deleteCnt < inputFiles.length)
 
     val fs = systemPath.getFileSystem(new Configuration)
     for (i <- 0 until appendCnt) {
       val sourcePath = new Path(inputFiles(i))
       val destPath = new Path(inputFiles(i) + ".copy")
       fs.copyToLocalFile(sourcePath, destPath)
     }
+    for (i <- 1 to deleteCnt) {
```
Comment (imback82) [inline src/test/scala/com/microsoft/hyperspace/index/HybridScanTest.scala:67]: can we be consistent of using `until` or `to` (just check above)

File: src/test/scala/com/microsoft/hyperspace/index/HybridScanTest.scala [hunk 12]
```diff
@@ -291,4 +340,339 @@ class HybridScanTest extends QueryTest with HyperspaceSuite {
...
+    }
+  }
+
+  test(
+    "Delete-only: filter index & parquet, json format, " +
+      "index relation should have additional filter for deleted files") {
+
+    Seq(
+      (sampleParquetDataLocationDelete, "index2", "parquet"),
+      (sampleJsonDataLocationDelete, "index5", "json")) foreach {
```
Comment (imback82) [inline src/test/scala/com/microsoft/hyperspace/index/HybridScanTest.scala:373]: There are so many indexes created for this suite that it's very hard to follow.

Can you at least describe what `index2` and `index5` are at the beginning of test so that I don't have to go back and forth to understand these indexes?

I am also fine creating the `index5` in this test since it's not being shared anyway (unless you are testing a scenario with multiple indexes in place).

File: src/main/scala/com/microsoft/hyperspace/index/rules/RuleUtils.scala (hunk 5 above)
Comment (sezruby) [reply src/main/scala/com/microsoft/hyperspace/index/rules/RuleUtils.scala:228]: The transformed logical relation is matched again with transformDown which causes stack overflow. I'll update the comment.

File: src/main/scala/com/microsoft/hyperspace/index/rules/RuleUtils.scala (hunk 9 above)
Comment (sezruby) [reply src/main/scala/com/microsoft/hyperspace/index/rules/RuleUtils.scala:292]: `l` from lineage. How about `lineageAttr` ??

File: src/main/scala/com/microsoft/hyperspace/index/rules/RuleUtils.scala (hunk 8 above)
Comment (sezruby) [reply src/main/scala/com/microsoft/hyperspace/index/rules/RuleUtils.scala:271]: I moved the condition in filter:
```
val newSchema = StructType(index.schema.filter(s =>
            baseRelation.schema.contains(s) || (filesDeleted.nonEmpty && s.name.equals(
              IndexConstants.DATA_FILE_NAME_COLUMN))))
```

File: src/test/scala/com/microsoft/hyperspace/index/HybridScanTest.scala (hunk 11 above)
Comment (sezruby) [reply src/test/scala/com/microsoft/hyperspace/index/HybridScanTest.scala:67]: These are intended not to use `i - 1` or `deletedCnt +1`. Which do you think is better here?

File: src/main/scala/com/microsoft/hyperspace/index/rules/RuleUtils.scala (hunk 5 above)
Comment (imback82) [reply src/main/scala/com/microsoft/hyperspace/index/rules/RuleUtils.scala:228]: Ah OK, makes sense. Thanks.

File: src/test/scala/com/microsoft/hyperspace/index/HybridScanTest.scala (hunk 11 above)
Comment (imback82) [reply src/test/scala/com/microsoft/hyperspace/index/HybridScanTest.scala:67]: Ah OK, then it should be fine. Let's add the delete behavior to the `def` comment.

File: src/main/scala/com/microsoft/hyperspace/index/IndexConstants.scala [hunk 13]
```diff
@@ -29,6 +29,8 @@ object IndexConstants {
 
   val INDEX_HYBRID_SCAN_ENABLED = "spark.hyperspace.index.hybridscan.enabled"
   val INDEX_HYBRID_SCAN_ENABLED_DEFAULT = "false"
+  val INDEX_HYBRID_SCAN_DELETE_ENABLED = "spark.hyperspace.index.hybridscan.delete.enabled"
```
[the remark this answers is not in the document]
Comment (sezruby) [reply src/main/scala/com/microsoft/hyperspace/index/IndexConstants.scala:32]: I think it's better to move these config related vals to HyperspaceConf and make some comments there. What do you think?
Can we do this with a separate PR?

File: src/main/scala/com/microsoft/hyperspace/index/rules/RuleUtils.scala (hunk 5 above)
Comment (sezruby) [reply src/main/scala/com/microsoft/hyperspace/index/rules/RuleUtils.scala]: > And I think we can call verifyPrerequisite from transformPlanToUseIndex since its comment also has this as a pre-requisite.

I added the assert stmt not function in `transformPlanToUseIndex`. Let's create the validate function later when we have another condition.

File: src/test/scala/com/microsoft/hyperspace/index/HybridScanTest.scala (hunk 12 above)
Comment (sezruby) [reply src/test/scala/com/microsoft/hyperspace/index/HybridScanTest.scala]: I renamed the indexes more descriptively. Let me know if they are still unclear. Thanks!

File: src/main/scala/com/microsoft/hyperspace/index/rules/RuleUtils.scala [hunk 14]
```diff
@@ -76,8 +79,12 @@ object RuleUtils {
       val commonCnt = inputSourceFiles.count(entry.allSourceFileInfos.contains)
       val deletedCnt = entry.allSourceFileInfos.size - commonCnt
```
Comment (apoorvedave1) [inline src/main/scala/com/microsoft/hyperspace/index/rules/RuleUtils.scala:80]: can move this in the else block below, so it's only calculated when delete enabled

File: src/main/scala/com/microsoft/hyperspace/index/rules/RuleUtils.scala [hunk 15]
```diff
@@ -216,25 +226,46 @@ object RuleUtils {
...
+
+        val (filesDeleted, filesAppended) =
+          if (HyperspaceConf.hybridScanDeleteEnabled(spark) && index.hasLineageColumn(spark)) {
+            val (exist, nonExist) = curFileSet.partition(index.allSourceFileInfos.contains)
+            val filesAppended = nonExist.map(f => new Path(f.name))
+            if (exist.length < index.allSourceFileInfos.size) {
+              (index.allSourceFileInfos -- exist, filesAppended)
+            } else {
+              (Nil, filesAppended)
+            }
```
Comment (apoorvedave1) [inline src/main/scala/com/microsoft/hyperspace/index/rules/RuleUtils.scala:249]: can we replace this section with just 
```scala
(index.allSourceFileInfos -- exist, filesAppended)
```

File: src/main/scala/com/microsoft/hyperspace/index/rules/RuleUtils.scala [hunk 16]
```diff
@@ -216,25 +226,46 @@ object RuleUtils {
     var unhandledAppendedFiles: Seq[Path] = Nil
```
Comment (apoorvedave1) [inline src/main/scala/com/microsoft/hyperspace/index/rules/RuleUtils.scala:226]: could we please move this variable closer to its usage? Also please add a description for this variable next to declaration?

File: src/main/scala/com/microsoft/hyperspace/index/IndexConstants.scala (hunk 13 above)
[the remark this answers is not in the document]
Comment (imback82) [reply src/main/scala/com/microsoft/hyperspace/index/IndexConstants.scala:32]: We can do the move as a separate PR, but let's add some comments in this PR.

**Comment by imback82** [comment]
@sezruby Can you resolve the conflict?

**Review by imback82 — COMMENTED** [review]
Few minor comments, but LGTM!

**Review by imback82 — APPROVED** [review]

**Comment by imback82** [comment]
Thanks for working on this @sezruby!
